mm2src/coins/nft/storage/wasm/wasm_storage.rs

```diff
@@ -28,144 +31,529 @@ impl IndexedDbNftStorage {
+        let table = db_transaction.table::<NftTxHistoryTable>().await?;
+        let mut txs = Vec::new();
+        for chain in chains {
+            let tx_tables = table
+                .get_items("chain", chain.to_string())
+                .await?
+                .into_iter()
+                .map(|(_item_id, tx)| tx);
+            let filtered = Self::take_txs_according_to_filters(tx_tables, filters)?;
+            txs.extend(filtered);
+        }
+        Self::take_txs_according_to_paging_opts(txs, max, limit, page_number)
     }
 
-    async fn add_txs_to_history<I>(&self, _chain: &Chain, _txs: I) -> MmResult<(), Self::Error>
+    async fn add_txs_to_history<I>(&self, _chain: &Chain, txs: I) -> MmResult<(), Self::Error>
     where
         I: IntoIterator<Item = NftTransferHistory> + Send + 'static,
         I::IntoIter: Send,
     {
-        todo!()
+        let locked_db = self.lock_db().await?;
+        let db_transaction = locked_db.get_inner().transaction().await?;
+        let table = db_transaction.table::<NftTxHistoryTable>().await?;
+        for tx in txs {
+            let tx_item = NftTxHistoryTable::from_tx_history(&tx)?;
+            table.add_item(&tx_item).await?;
+        }
+        Ok(())
     }
 
-    async fn get_last_block_number(&self, _chain: &Chain) -> MmResult<Option<u32>, Self::Error> { todo!() }
+    async fn get_last_block_number(&self, chain: &Chain) -> MmResult<Option<u64>, Self::Error> {
+        let locked_db = self.lock_db().await?;
+        let db_transaction = locked_db.get_inner().transaction().await?;
+        let table = db_transaction.table::<NftTxHistoryTable>().await?;
+        get_last_block_from_table(chain, table, NftTxHistoryTable::CHAIN_BLOCK_NUMBER_INDEX).await
+    }
 
     async fn get_txs_from_block(
         &self,
-        _chain: &Chain,
-        _from_block: u32,
+        chain: &Chain,
+        from_block: u64,
     ) -> MmResult<Vec<NftTransferHistory>, Self::Error> {
-        todo!()
+        let locked_db = self.lock_db().await?;
+        let db_transaction = locked_db.get_inner().transaction().await?;
+        let table = db_transaction.table::<NftTxHistoryTable>().await?;
+        let mut cursor = table
+            .cursor_builder()
+            .only("chain", chain.to_string())
+            .map_err(|e| WasmNftCacheError::GetLastNftBlockError(e.to_string()))?
+            .bound("block_number", BeBigUint::from(from_block), BeBigUint::from(u64::MAX))
+            .open_cursor(NftTxHistoryTable::CHAIN_BLOCK_NUMBER_INDEX)
+            .await
+            .map_err(|e| WasmNftCacheError::GetLastNftBlockError(e.to_string()))?;
+
+        let mut res = Vec::new();
+        while let Some((_item_id, item)) = cursor
+            .next()
+            .await
+            .map_err(|e| WasmNftCacheError::GetLastNftBlockError(e.to_string()))?
+        {
+            let tx = tx_details_from_item(item)?;
+            res.push(tx);
+        }
+        drop_mutability!(res);
+        Ok(res)
```

Comment:
here `drop_mutability` doesn't seem too necessary IMO, you just returning the value after it anyway

Reply:
ok done